File: python/numeric.py
# Implementation of numeric algorithms

def itoa(v):
    value = v
    while True :
        yield value
        value += 1

def accumulate(seq, init, func):
    return reduce(func, seq, init)

def inner_production(seq1, seq2, init):
    if len(seq1) != len(seq2) :
        raise TypeError, "second sequence should be as long as the first one"
    return reduce(lambda x, y: x + y[0] * y[1], zip(seq1, seq2), init)

def adjacent_difference(seq):
    res = [seq[0]]
    def diff(x, y):
        res.append(y - x)
        return y
    reduce(diff, seq)
    return res

def partial_sum(seq):
    res = [seq[0]]
    def sum(x, y):
        res.append(x + y)
        return x + y
    reduce(sum, seq)
    return res

File: lua/numeric.lua
function itoa(v)
    local value = v
    return function()
        value = value + 1
        return value - 1
    end
end

function accumulate(seq, init, func)
    local sum = init
    for i = 1, #seq do
        sum = func(sum, seq[i])
    end
    return sum
end

-- require #s1 == #s2
function innerProduction(s1, s2, init)
    local sum = init
    for i = 1, #s1 do
        sum = sum + s1[i] * s2[i]
    end
    return sum
end

function adjacentDifference(seq)
    if #seq == 0 then return {} end
    local res = {seq[1]}
    for i = 2, #seq do
        res[i] = seq[i] - seq[i-1]
    end
    return res
end

function partialSum(seq)
    if #seq == 0 then return {} end
    local res = {seq[1]}
    for i = 2, #seq do
        res[i] = res[i-1] + seq[i]
    end
    return res
end

return {
    itoa = itoa,
    accumulate = accumulate,
    innerProduction = innerProduction,
    adjacentDifference = adjacentDifference,
    partialSum = partialSum
}
